### PR TITLE
refactor(core): move actions inside core

### PR DESCRIPTION
**Pull request by ematipico**
## Changes

This PR moves actions inside core. Actions are now part of the pipeline, and they are loaded in the same way the middleware is loaded:
- using `ModuleLoader` in development
- using its entry point after the first build
- using its `moduleId` when build for SSR

The internal virtual module of the actions is in charge of emitting a stub of the actions in case the user doesn't have any.

The injections of the actions during development is done inside the `route.ts`. I decided to do it there because I wanted to assume that the actions must be already loaded when creating the `RenderContext`, which is our agnostic runtime renderer. Because fo that, the `Pipeline` class also exposes a method called `setActions`.

## Testing

Current tests and functionality should stay the same. The Actions middleware was already set to be the last one, which is the same as we just moved inside the `lastNext` callback we have in core.

## Docs

N/A

### CODE REVIEW

**Review by florian-lefebvre — DISMISSED** [review]
Although this is a refactor, maybe should we have a changeset to be safe in case people face issues

**Comment by ematipico** [comment]
Turns out it is not as simple as that, it requires a full refactor of the architecture. I'll see to that

**Comment by ematipico** [comment]
I checked the e2e test that failed, and the issue isn't much clear. The db package fails to run the seed file, however it doesn't happen when I run the dev server for the fixture.

Also, when running the fixture locally, the likes action (the action) works as expected, but the query doesn't return the expected data. This doesn't happen for the comments action. 

As for now, I commented on them.

File: packages/astro/e2e/fixtures/actions-blog/src/actions/index.ts [hunk 1]
```diff
@@ -15,16 +15,16 @@ export const server = {
 			handler: async ({ postId }) => {
 				await new Promise((r) => setTimeout(r, 500));
 
-				const { likes } = await db
+				const result  = await db
 					.update(Likes)
 					.set({
-						likes: sql`likes + 1`,
+						likes: sql`${Likes.likes} + 1`,
```
Comment (bholmesdev) [inline packages/astro/e2e/fixtures/actions-blog/src/actions/index.ts:21]: Hm, this doesn't seem right. Any reason we stopped incrementing from sql for this test?

File: packages/astro/e2e/fixtures/actions-blog/src/components/PostComment.tsx [hunk 2]
```diff
@@ -17,7 +18,7 @@ export function PostComment({
 			<form
 				method="POST"
 				data-testid="client"
-				action={actions.blog.comment}
+				action={actions.blog.comment.toString()}
```
Comment (bholmesdev) [inline packages/astro/e2e/fixtures/actions-blog/src/components/PostComment.tsx:21]: Just noting that if this is required, it's a user-side bug. Understand if this is just a debug in-progress

File: packages/astro/src/core/render-context.ts [hunk 3]
```diff
@@ -192,6 +194,15 @@ export class RenderContext {
 			}
 			let response: Response;
 
+			if (!ctx.isPrerendered) {
```
Comment (bholmesdev) [inline packages/astro/src/core/render-context.ts:197]: Nice, this looks like the fix we needed.

And I'll add some context from @JacobNWolf 's issue to make sure we test it: Actions were being bundled into edge middleware, when they should only be bundled into your server runtime. It looks like this moves the _functionality_ into the server runtime, but I wanna make sure we make a test project (ideally a fixture) that confirms actions aren't bundled into middleware by accident still.

Comment (JacobNWolf) [reply packages/astro/src/core/render-context.ts:197]: Amazing. Thank you guys.

File: packages/astro/e2e/fixtures/actions-blog/src/components/PostComment.tsx (hunk 2 above)
Comment (ematipico) [reply packages/astro/e2e/fixtures/actions-blog/src/components/PostComment.tsx:21]: That's intentional. For some reason, the `toString()` isn't called anymore. Has something changed in the TSX compiler? I bet that if we use a template literal, the `toString()` function is called, however... like this, `toString()` isn't automatically called, so I am afraid the transformation is doing something fishy

**Comment by ematipico** [comment]
@JacobNWolf, can you please use the latest preview release and see if that fixes your issue?

File: packages/astro/test/fixtures/actions/src/pages/invalid.astro [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ---
 import { actions } from "astro:actions";
-
+console.log(actions.subcribe)
```
Comment (matthewp) [inline packages/astro/test/fixtures/actions/src/pages/invalid.astro:3]: left in for debugging i presume, should remove

File: packages/astro/src/actions/loadActions.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * It accepts a module loader and the astro settings, and it attempts to load the middlewares defined in the configuration.
+ *
+ * If not middlewares were not set, the function returns an empty array.
+ */
+export async function loadActions(moduleLoader: ModuleLoader) {
+	try {
+		return (await moduleLoader.import(ASTRO_ACTIONS_INTERNAL_MODULE_ID)) as SSRActions;
+	} catch (error: any) {
+		// TODO create astro error
```
Comment (florian-lefebvre) [inline packages/astro/src/actions/loadActions.ts:14]: should this be addressed in this PR?

File: packages/astro/src/actions/runtime/virtual/shared.ts [hunk 6]
```diff
@@ -309,3 +309,7 @@ const actionResultErrorStack = (function actionResultErrorStackFn() {
 		},
 	};
 })();
+
+export function astroCalledServerError(): AstroError {
+	return new AstroError(ActionCalledFromServerError);
+}
```
Comment (florian-lefebvre) [inline packages/astro/src/actions/runtime/virtual/shared.ts:315]: Is this needed? AFAIK we usually directly throw an astro error without wrapping it in a function

File: packages/astro/src/vite-plugin-astro-server/plugin.ts [hunk 7]
```diff
@@ -25,6 +25,7 @@ import { DevPipeline } from './pipeline.js';
 import { handleRequest } from './request.js';
 import { setRouteError } from './server-state.js';
 import { trailingSlashMiddleware } from './trailing-slash.js';
+import { loadActions } from '../actions/loadActions.js';
```
Comment (florian-lefebvre) [inline packages/astro/src/vite-plugin-astro-server/plugin.ts:28]: import is unused

File: packages/astro/src/vite-plugin-astro-server/pipeline.ts [hunk 8]
```diff
@@ -29,6 +29,7 @@ export class DevPipeline extends Pipeline {
 	// renderers are loaded on every request,
 	// so it needs to be mutable here unlike in other environments
 	override renderers = new Array<SSRLoadedRenderer>();
+	// override actions = undefined;
```
Comment (florian-lefebvre) [inline packages/astro/src/vite-plugin-astro-server/pipeline.ts:32]: why is this commented?

File: packages/astro/src/vite-plugin-astro-server/pipeline.ts [hunk 9]
```diff
@@ -212,16 +213,4 @@ export class DevPipeline extends Pipeline {
 	setManifestData(manifestData: RoutesList) {
 		this.routesList = manifestData;
 	}
-
-	rewriteKnownRoute(route: string, sourceRoute: RouteData): ComponentInstance {
```
Comment (florian-lefebvre) [inline packages/astro/src/vite-plugin-astro-server/pipeline.ts:216]: Why is this removed?

File: packages/astro/src/actions/runtime/virtual/shared.ts (hunk 6 above)
Comment (ematipico) [reply packages/astro/src/actions/runtime/virtual/shared.ts:315]: We do, however this particular function is used inside the template file `astro.mjs`, where we only import `astro:actions` virtual module.

**Review by florian-lefebvre — APPROVED** [review]

**Review by bholmesdev — APPROVED** [review]
Thanks for taking `action` on this!
